dialogue-apache-hc5-client/src/main/java/com/palantir/dialogue/hc5/DialogueManagedHttpClientConnection.java

```diff
@@ -0,0 +1,174 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.hc5;
+
+import java.io.IOException;
+import java.io.InterruptedIOException;
+import java.net.Socket;
+import java.nio.charset.CharsetDecoder;
+import java.nio.charset.CharsetEncoder;
+import java.util.concurrent.atomic.AtomicBoolean;
+import javax.annotation.Nullable;
+import javax.net.ssl.SSLSession;
+import javax.net.ssl.SSLSocket;
+import org.apache.hc.client5.http.io.ManagedHttpClientConnection;
+import org.apache.hc.core5.http.ClassicHttpRequest;
+import org.apache.hc.core5.http.ClassicHttpResponse;
+import org.apache.hc.core5.http.ContentLengthStrategy;
+import org.apache.hc.core5.http.HttpEntity;
+import org.apache.hc.core5.http.HttpException;
+import org.apache.hc.core5.http.config.Http1Config;
+import org.apache.hc.core5.http.impl.io.DefaultBHttpClientConnection;
+import org.apache.hc.core5.http.impl.io.SocketHolder;
+import org.apache.hc.core5.http.io.HttpMessageParserFactory;
+import org.apache.hc.core5.http.io.HttpMessageWriterFactory;
+import org.apache.hc.core5.io.CloseMode;
+import org.apache.hc.core5.util.Identifiable;
+import org.apache.hc.core5.util.Timeout;
+
+/**
+ * Connection which produces behavior matching the default managed connection from httpcore while
+ * providing the functionality proposed in
+ * <a href="https://issues.apache.org/jira/browse/HTTPCORE-639">HTTPCORE-639</a>.
+ */
+final class DialogueManagedHttpClientConnection extends DefaultBHttpClientConnection
+        implements ManagedHttpClientConnection, Identifiable {
+
+    private final String id;
+    private final AtomicBoolean closed = new AtomicBoolean();
+    private final AtomicBoolean availableDataCheckDisabled = new AtomicBoolean();
+
+    @Nullable
+    private Timeout socketTimeout;
+
+    DialogueManagedHttpClientConnection(
+            String id,
+            @Nullable CharsetDecoder charDecoder,
+            @Nullable CharsetEncoder charEncoder,
+            Http1Config h1Config,
+            ContentLengthStrategy incomingContentStrategy,
+            ContentLengthStrategy outgoingContentStrategy,
+            HttpMessageWriterFactory<ClassicHttpRequest> requestWriterFactory,
+            HttpMessageParserFactory<ClassicHttpResponse> responseParserFactory) {
+        super(
+                h1Config,
+                charDecoder,
+                charEncoder,
+                incomingContentStrategy,
+                outgoingContentStrategy,
+                requestWriterFactory,
+                responseParserFactory);
+        this.id = id;
+    }
+
+    @Override
+    public String getId() {
+        return id;
+    }
+
+    @Override
+    public void bind(SocketHolder socketHolder) throws IOException {
```

Comment:
I think I'd find it reassuring to have little comments or something on these methods:

```
/** Identical to {@link DefaultManagedHttpClientConnection#bind}. */
```

So that readers/maintainers can clearly identify which are the bits we've added